drag and drop

porridgewithraisins opened the pull request.

porridgewithraisins committed:
add drag and drop functionality

adds a drag data get handler, and adds the file:// URIs of the current
selection to it. there are still some rough edges.

porridgewithraisins committed:
add ccls cache to gitignore

porridgewithraisins committed:
change editorconfig rule for git commit message

recommended column length for commit message is 72 chars.

ladybunne commented:
This would be a really nice feature to have, and it looks like the PR is all ready to go - is there anything stopping it from being merged?

porridgewithraisins committed:
make drag and drop proper

see comment in below link

https://github.com/cboxdoerfer/fsearch/issues/11#issuecomment-2892485422

porridgewithraisins committed:
fix memory leak of GString

g_string_free needs to be used for GStrings, g_free was used earlier

porridgewithraisins commented:
@ladybunne please check #11 for the discusssion (its a really old issue so scroll to the end)

porridgewithraisins committed:
remove unnecessary variables and logs

cheesetosti reviewed (approved):
Works fine. Only missing any visual cursor feedback (e.g. plus sign indicator next to cursor to indicate holding a file with the mouse cursor) but other than that it works flawlessly!

cboxdoerfer commented:
Awesome, thanks for the PR. I'm testing it right now and will be reporting back to you!

cboxdoerfer commented on src/fsearch_list_view.c:
```diff
@@ -1110,6 +1123,49 @@ on_fsearch_list_view_bin_drag_gesture_begin(GtkGestureDrag *gesture,
+        uri_array[index] = NULL;
+        gtk_selection_data_set_uris(selection_data, uri_array);
+        g_list_free_full(uris, g_free);
+        for (int i = 0; uri_array[i] != NULL; i++) {
```
you can use `g_clear_pointer(&uri_array, g_strfreev);` here instead to clean up the whole string array.

cboxdoerfer commented on src/fsearch_window.c:
```diff
@@ -725,6 +725,36 @@ on_listview_row_is_selected(int row, gpointer user_data) {
 }
 
+static GList*
+listview_on_drag_data_get(GList *indices, gpointer user_data) {
```
You don't need to pass the indices of all the selected entries here. Instead you can simply call `db_view_selection_for_each(db_view, some_prepend_path_helper, &uris);` here to collect all the selected paths.

cboxdoerfer replied:
Or even better, instead of having `some_prepend_path_helper()` building a new GList, use it to build a string array right away, as that's what we need for `gtk_selection_data_set_uris()`.

porridgewithraisins committed:
Address review comments

1. Use db_..._for_each instead of getting indices and passing that
   and...
2. Use g_clear_pointer
3. Assorted formatting and style fixes

cboxdoerfer commented on src/fsearch_window.c:
```diff
@@ -725,6 +725,62 @@ on_listview_row_is_selected(int row, gpointer user_data) {
+        return NULL;
+    }
+
+    GListAccumulator acc = { NULL };
```
You don't need to wrap `GList` with `GListAccumulator`. You can pass the list instead like so:

```
GList *uri_list = NULL;
db_view_selection_for_each(db_view, map_to_uri_helper, &uri_list);
```

and then use it in `map_to_uri_helper`:

```
GList **list = user_data; 
*list = g_list_prepend(*list, uri);
```

porridgewithraisins replied:
Ah yes silly me.

cboxdoerfer commented on src/fsearch_window.c:
```diff
@@ -725,6 +725,62 @@ on_listview_row_is_selected(int row, gpointer user_data) {
+
+    gchar **uri_array = g_new(gchar *, g_list_length(acc.list) + 1);
+    int i = 0;
+    for (GList *l = g_list_last(acc.list); l != NULL; l = l->prev) {
```
I suppose you're walking the list from the last to the first element to get the order right? That's fine in theory, but just be aware that currently `db_view_selection_for_each()` has no defined order in which it walks through the selected entries, as the selection is stored in a hash table. So the order will likely be messed up either way.

cboxdoerfer replied:
Still, the code is fine as it is. We really should at least try to get the order right, in case we switch data structures for selections.

porridgewithraisins committed:
remove g_list_last() call, populate array backwards

remove useless accumulator as well

porridgewithraisins commented:
@cboxdoerfer this is how multiple item dragging works right now:

https://github.com/user-attachments/assets/39167d57-484b-4a3a-b7a7-48d882601f79


I suppose, like you said, what we should do is make it so that click selects only on release? that way the drag event will take precedence unless they release?

porridgewithraisins committed:
add clarifying comment for uri_array_len

vittorioromeo commented:
@cboxdoerfer any reason why this was never merged? I miss this feature badly from Everything

derickso commented:
I just applied https://github.com/cboxdoerfer/fsearch/pull/640 to the latest mainline branch on CachyOS and confirmed drag and drop was working. Claude says the newer pull request is more robust and matches latest code more cleanly.
